lang/java/avro/src/main/java/org/apache/avro/Protocol.java

```diff
@@ -275,15 +274,29 @@ public Protocol(Protocol p) {
 
   public Protocol(String name, String doc, String namespace) {
     super(PROTOCOL_RESERVED);
-    this.name = name;
+    setName(name, namespace);
     this.doc = doc;
-    this.namespace = namespace;
   }
 
   public Protocol(String name, String namespace) {
     this(name, null, namespace);
   }
 
+  private void setName(String name, String namespace) {
+    int lastDot = name.lastIndexOf('.');
+    if (lastDot < 0) {
+      this.name = name;
+      this.namespace = namespace;
+    } else {
```

Comment:
What if namespace is "java" and name is "lang.String" 
Here, a suggestion :
```java
        StringBuilder namespaceBuilder = new StringBuilder((namespace != null) ? namespace : "");
        if (namespaceBuilder.length() > 0 && namespaceBuilder.charAt(namespaceBuilder.length() - 1) != '.') {
            namespaceBuilder.append('.');
        }
        namespaceBuilder.append(name.substring(0, lastDot));
       this.namespace = namespaceBuilder.toString();
```

Reply:
The intent is to use the same namespace the the definition itself, and any enclosed definitions that don't specify one themselves. This is a lot easier to comprehend than using different namespaces.